### PR TITLE
managed-agent will periodically call dispatcher.Dispatch

### LINKED ISSUE #2343: Fleet stuck in "Updating..." status when attempting to upgrade agents to v8.6.1 or v8.6.2
# Issue

Since upgrading from Elastic Cloud 8.5.3, I have been unable to manage Elastic Agent integrations through Fleet in Kibana.

I initially upgraded from 8.5.3->8.6.1 and scheduled an Agent upgrade. A few of the agents upgraded as expected, the remaining remained stuck in "Updating..." status in Fleet for more than a week. I later updated to 8.6.2 and continued to experience the same issue.

I cancelled out the "Updating..." status per the instructions [in this post](https://discuss.elastic.co/t/fleet-server-status-stuck-at-updating-but-seems-to-be-working-normally/318193/2), which returned the agents to Healthy status, as promised. A second attempt at upgrading yielded the same results. After cancelling out the "Updating..." status again, I determined that I was able to schedule agents to upgrade immediately and it would usually succeed, as long as they were in "Healthy" status, but if an "Offline" or even a "Not Healthy" agent was included, the upgrade would fail. If I attempted to schedule the update for a later time or with a maintenance window other than "immediately" the upgrade fails.

Additionally, I observed this morning when I checked the "Agent Activity" log in fleet that multiple activities showing "Reassigning X of Y agents to [new policy]" that I executed approximately a week ago are still showing as in progress. However, in this case, when I investigate, all agents have been reassigned to the new policy and are reporting in as expected. I don't know whether the two items are related.

# Deployment

My deployment is:

  * Elastic Cloud 8.6.2
  * Fleet-Managed Elastic Agent in both Windows and Linux
  * Discussion Forum: https://discuss.elastic.co/t/elastic-agents-fail-to-upgrade-from-fleet-shows-updating-status-for-more-than-a-week/325426/12

# Steps to Reproduce

  1. Upgrade Elastic Cloud to 8.6.1 or 8.6.2
  2. Schedule an agent upgrade in Fleet to occur at a specific time.

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Add a timer in the goroutine that passes actions from fleet-gateway to the dispatcher that calls dispatch with no actions.

## Why is it important?

Scheduled actions are only ran when the `dispatch.Dispatch()` method is called.
Currently this method is only called when an action list is received from the fleet-gateway.

If the gateway does not send any actions after a checkin, the dispatcher is not called and scheduled actions are not checked (until a new action is received).

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

## How to test this PR locally

- Schedule an upgrade for a future time
- upgrade should run if no new action is sent after scheduled time

## Related issues

- Closes #2343

### CODE REVIEW

**Comment by ycombinator** [comment]
While this change will have the desired effect of scheduled actions being dequed and dispatched even when no subsequent actions are received from Fleet, I have two concerns/comments:
* With the proposed approach, we will call the [`dispatch.Dispatch()` method](https://github.com/elastic/elastic-agent/blob/2c41221cd2fb33bf410f52fc5f364316d902c350/internal/pkg/agent/application/dispatcher/dispatcher.go#L104) once every 1 - 1.5 seconds by default, since that's the [default configuration](https://github.com/elastic/elastic-agent/blob/2c41221cd2fb33bf410f52fc5f364316d902c350/internal/pkg/agent/application/gateway/fleet/fleet_gateway.go#L37-L38) for how often the agent is scheduled to checkin with Fleet Server.  By comparison, currently we are calling `dispatch.Dispatch()` much less frequently since most of the times there will likely not be any actions being returned in the checkin response from the Fleet Server.  I'm not sure if this change will have any impact on performance, particularly CPU usage?
* I think the architecture of calling `dispatch.Dispatch()` only when we have a _successful checkin call_ to the Fleet Server might be exposing a more fundamental issue in our design, particularly in relation to scheduled actions.  What if the Fleet Server is temporarily unreachable for some reason?  In the current design, we will not execute any scheduled actions until the connection is restored and we have a successful checkin call again.  In fact, because scheduled actions [support expiration](https://github.com/elastic/elastic-agent/blob/2c41221cd2fb33bf410f52fc5f364316d902c350/internal/pkg/fleetapi/action.go#L60-L61), we may _never_ execute a scheduled action if the connection is restored only after the expiration period has elapsed.  I don't know enough about the history of the project to know if this behavior is desired or not.  Today the only scheduled action possible is upgrade; I would wager that upgrades should be processed even if there's some problem in connecting to the Fleet Server for the checkin call but maybe it's by design we don't want to do this?

An alternative architecture might be for the `managedConfigManager.Run()` method to start the dispatcher in its own goroutine so it can keep (efficiently) processing it's priority queue over time.  And then we can keep the current approach of sending actions from the fleet gateway to the dispatcher only when the checkin call is successful and returns a non-empty list of actions.

**Comment by michel-laterman** [comment]
@ycombinator
I thought the checkin was a long-poll checkin, however you raised a good point about connectivity. I'll use an alternate mechanism instead.

File: internal/pkg/agent/application/managed_mode.go [hunk 1]
```diff
@@ -209,12 +212,17 @@ func (m *managedConfigManager) Run(ctx context.Context) error {
...
 	// pass actions collected from gateway to dispatcher
 	go func() {
+		t := time.NewTimer(dispatchFlushInterval)
 		for {
 			select {
 			case <-ctx.Done():
 				return
+			case <-t.C: // periodically call the dispatcher to handle scheduled actions.
+				m.dispatcher.Dispatch(ctx, actionAcker)
+				t.Reset(dispatchFlushInterval)
```
Comment (ycombinator) [inline internal/pkg/agent/application/managed_mode.go:222]: 👍.  I think this is a slightly better approach because it at least decouples the processing of the dispatcher queues from the Fleet Checkin call.

Comment (ycombinator) [reply internal/pkg/agent/application/managed_mode.go:222]: I think we can have an even more efficient implementation of "when to re-dispatch" by deciding to process the action queue either when new actions are added to it (already handled by the `case` below this line) or just-in-time for the next scheduled action (by looking at the next scheduled action's time and setting up a `time.Timer` accordingly, rather than on a fixed interval like we are doing here).  But we can make this change in a separate PR if it's too complicated for this one.

**Review by ycombinator — APPROVED** [review]

File: internal/pkg/agent/application/managed_mode.go [hunk 2]
```diff
@@ -33,6 +33,9 @@ import (
 	"github.com/elastic/elastic-agent/pkg/core/logger"
 )
 
+// dispatchFlushInterval is the min time between calls to dispatcher.Dispatch
```
Comment (ycombinator) [inline internal/pkg/agent/application/managed_mode.go:36]: min or max?

Comment (michel-laterman) [reply internal/pkg/agent/application/managed_mode.go:36]: max, i'll fix the comment

**Comment by cmacknz** [comment]
> I thought the checkin was a long-poll checkin

Just to confirm, the check in requests do use long polling but fleet-server is responsible for holding the request open. The short 1 - 1.5s request durations on the agent only come into play when the server has terminated the long poll request.

Big thanks to @ycombinator for suggesting that we decouple the execution of scheduled actions from the checkin period regardless. This is a much better approach and avoids the possibility of hard to find side effects as we start increasing the checkin duration significantly for scalability reasons (from 5 to 30 minutes).

File: changelog/fragments/1677797191-dispatch-periodic-flush.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: bug-fix
+
+# Change summary; a 80ish characters long description of the change.
+summary: Periodically run the dispatcher from a timer.
```
Comment (cmacknz) [inline changelog/fragments/1677797191-dispatch-periodic-flush.yaml:14]: Can you rephrase this to describe the impact to the user? What bug does this fix? What features are effected?

The descriptions here are good but they will only really make sense to engineers, the change log is for users.

Comment (cmacknz) [reply changelog/fragments/1677797191-dispatch-periodic-flush.yaml:14]: None of our users know what the dispatcher is or does.

File: internal/pkg/agent/application/managed_mode.go [hunk 4]
```diff
@@ -209,12 +212,17 @@ func (m *managedConfigManager) Run(ctx context.Context) error {
 
 	// pass actions collected from gateway to dispatcher
 	go func() {
+		t := time.NewTimer(dispatchFlushInterval)
 		for {
 			select {
 			case <-ctx.Done():
 				return
+			case <-t.C: // periodically call the dispatcher to handle scheduled actions.
```
Comment (cmacknz) [inline internal/pkg/agent/application/managed_mode.go:220]: You need to add a test for this, preferably one that catches the original bug you are fixing here.

File: changelog/fragments/1677797191-dispatch-periodic-flush.yaml (hunk 3 above)
Comment (cmacknz) [reply changelog/fragments/1677797191-dispatch-periodic-flush.yaml]: That is an excellent changelog entry, thanks!
